### PR TITLE
misc: add Makefile and add pyright to requirements-optional.txt

### PR DESCRIPTION
**Pull request by superlopuh**
This should make it easier for new contributors to get set up. I couldn't figure out how to get the venv part in there, I just require venv to be activated for the pip commands.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# default target, when make executed without arguments
+all: venv
+
+install: FORCE
+	pip install --upgrade pip
+	pip install -r requirements.txt --require-venv
+	pip install -r requirements-optional.txt --require-venv
+	pip install --editable . --require-venv
+
+requirements.txt:
```
Comment (AntonLydike) [inline Makefile:13]: I don't understand how this rule "makes" the `requirements.txt`

What#s the purpose of this rule?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# define the name of the virtual environment directory
+VENV := venv
+
+# default target, when make executed without arguments
+all: venv
```
Comment (AntonLydike) [inline Makefile:5]: Why do we have no rule to make `venv`, but it's the default target?

I'd suggest a rule like:

```
venv:
  python3 -m venv ${VENV}
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+install: FORCE
```
Comment (AntonLydike) [inline Makefile:7]: What is the purpose of `FORCE`

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	pyright
+
+lint-ci: pyright-ci.json
+	pyright -p pyright-ci.json
+
+clean:
+	rm -r $(VENV)
+	find . -type f -name '*.pyc' -delete
+
+.PHONY: all install update venv test clean
```
Comment (AntonLydike) [inline Makefile:38]: `venv` should probably not be phony, right? Am I missing something?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	# Executes filecheck tests
+	lit tests/filecheck
+
+format:
+	yapf -ir xdsl
+
+lint:
+	pyright
+
+lint-ci: pyright-ci.json
```
Comment (AntonLydike) [inline Makefile:31]: Why is `pyright-ci.json` a requirement here? Do we need to keep track of it?

**Review by AntonLydike — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (superlopuh) [reply Makefile:5]: yeah I'll add it back
- [ ] add it back

File: Makefile (hunk 3 above)
Comment (superlopuh) [reply Makefile:7]: It means that even if the dependencies haven't changed, `make force` will make it run. It circumvents the lazy/incremental running of make instructions

File: Makefile (hunk 4 above)
Comment (superlopuh) [reply Makefile:38]: To be honest I'm winging it with a fair bit of this makefile, would appreciate help on how to improve it. I copied an existing makefile from a tutorial and have just been adding the rules to the .PHONY line

File: Makefile (hunk 5 above)
Comment (superlopuh) [reply Makefile:31]: I guess it's not really useful to have it in the requirement slot

File: Makefile (hunk 4 above)
Comment (AntonLydike) [reply Makefile:38]: If you want we can try this together, I'm also not super familiar with how Makefiles work, but I'm sure we can figure out a nice way together!

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# define the name of the virtual environment directory
+VENV := venv
```
Comment (webmiche) [inline Makefile:2]: Maybe name it `xDSL` or something? Or does it have to be named venv?

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+test: update
+	# Executes pytests which are located in tests/
+	pytest
+	# Executes filecheck tests
+	lit tests/filecheck
+
+format:
+	yapf -ir xdsl
+
+lint:
```
Comment (webmiche) [inline Makefile:28]: can this be parametrized to only run it on one folder? (same for formatting, or even testing).

**Review by webmiche — APPROVED** [review]
This is extremely cool, thanks a lot!
Reading Makefile gives me PTSD, but I guess we can't get around that :wink:

**Comment by superlopuh** [comment]
Let's close this for now, and try again when we next hit this issue.

**Pull request closed without merging**